### PR TITLE
feat(maintenance): archive merged state.db sources on --apply so the doctor stops flagging them

### LINKED ISSUE #1390: feat(maintenance): archive merged state.db sources on --apply so the doctor stops flagging them
## What happens

`scripts/maintenance/merge-fragmented-state-dbs.js --apply` copies the rows from each `--source` database into the canonical CLI store (`~/.egc/egc/state.db`) and backs up the canonical file first. It never touches the sources. Right after a successful `--apply`, `egc doctor` still prints:

```
WARNING: found 2 stray state.db copies left behind by older versions:
  C:\Users\comp\.config\opencode\egc\state.db (184320 bytes, last write 2026-07-30T01:04:27.566Z)
  C:\Users\comp\.gemini\egc\state.db (684032 bytes, last write 2026-07-30T00:19:10.899Z)
```

The warning is technically true (the files are still there) but the user just did what the doctor asked and gets the same warning back. Reported in #1380 by @Akisolu, first person to run the merge script on real Windows state: dry run and `--apply` both worked, the doctor did not notice.

Leaving the copies in place has a second cost: `resolveStateStoreDbPath()` in `mcp/servers/egc-memory/src/index.ts` probes the known harness directories before falling back to `~/.egc`, so as long as `~/.gemini/egc/state.db` exists the memory server's `detect_patterns` reads events from the stale copy instead of the canonical store.

## Where

- `scripts/maintenance/merge-fragmented-state-dbs.js`: `mergeStateDbs` returns after `flush()` with the sources untouched.
- `scripts/doctor.js`: `findStateDbFragments` matches the exact filename `state.db` under `<harness>/egc/`, so a renamed source stops being flagged.

## Proposal (open to discussion)

After a successful `--apply`, rename each merged source to `state.db.merged-<timestamp>.bak` next to where it was (never delete, same rule as the canonical backup), and print the new names. A `--keep-sources` flag would opt out. The JSON output gains an `archived` array so scripts and tests can assert on it. The doctor hint from #1389 then closes the loop: run the printed command, add `--apply`, run `egc doctor`, clean.

## Notes

- The merge must stay dry-run by default and the rename must only happen on `--apply` after the transaction committed.
- Add a test that runs `--apply` against two temp sources and asserts both were renamed and that a fresh doctor scan finds no fragments.

### PR DESCRIPTION
**Pull request by Fmarzochi**
## What Changed

`scripts/maintenance/merge-fragmented-state-dbs.js` now finishes the job it starts. After a successful `--apply`, each merged source is renamed next to itself:

```
<source>/state.db  ->  <source>/state.db.merged-<timestamp>.bak
```

- SQLite sidecars that belong to the source (`-wal`, `-shm`, `-journal`) move with it and keep their suffix.
- Nothing is deleted, same rule as the canonical backup the script already takes before writing.
- The rename happens only after the canonical store is written, flushed and closed. A source that failed to merge (file not found) stays where it is. A rename that fails, for example a file held open by another process on Windows, is reported in that source's entry as `archiveError` instead of failing a merge that already succeeded.
- `--keep-sources` opts out. Dry runs never rename anything.
- The JSON result gains `keepSources` and `archived` (one `{ source, archivedTo }` per renamed file); each per-source report gains `archivedTo`.
- New guard: a `--source` that resolves to the canonical store is refused with `source is the canonical store`, so the live store can never be archived by mistake.
- The usage text lists the flag and says what `--apply` does to the sources.

Tests (11 in the suite, 5 new): archive with sidecar and byte-for-byte check, dry run and failed source leave files alone, canonical passed as source refused, the CLI with and without `--keep-sources`, and an integration case that seeds a home with two stray copies, runs `egc doctor` before (two strays listed) and after `--apply` (clean).

## Why This Change

`egc doctor` scans for the exact filename `state.db` under each tool's `egc/` folder. The merge script never touched its sources, so right after the person did what the doctor asked, the doctor printed the same warning again. Fixes #1390.

Leaving the copies in place also kept the memory server's harness probe (`resolveStateStoreDbPath` in `mcp/servers/egc-memory/src/index.ts`) pointing `detect_patterns` at the stale copy instead of the canonical store.

Pairs with #1391, which makes the doctor print a runnable command: run it, add `--apply`, run `egc doctor`, clean.

## Testing Done

- [x] Manual testing completed
- [x] Automated tests pass locally (`node tests/run-all.js`)
- [x] Edge cases considered and tested
- [ ] If this PR touches a file shared across concurrent EGC processes (encryption key, state files, install-state, lockfiles under `~/.egc/`), a concurrent-access test was added or updated (not applicable: the script is a manual one-off tool that renames files no session writes to anymore; a rename that loses the race is reported, never retried)

## Type of Change
- [x] `feat:` New feature

## Security & Quality Checklist
- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly
- [x] Shell scripts pass shellcheck (if applicable)
- [x] Pre-commit hooks pass locally (if configured)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## Documentation
- [x] Updated relevant documentation (script header and usage text)
- [x] Added comments for complex logic
- [ ] README updated (not needed)

## Credit

Reported by @Akisolu in #1380: the first run of the consolidation script on real Windows state, dry run and `--apply`, followed by a doctor that still listed both copies. The archive-and-never-delete shape is the one she proposed.

## Hardening after review

Four points raised in review, all taken:

- The canonical guard compares filesystem identity (`fs.realpathSync.native` on both sides, case-folded fallback on Windows when a path does not exist yet), so a symlink to the live store or the same file spelled in another case is refused too. Test: symlink on POSIX, upper-cased name on Windows.
- Every destination, main file and sidecars, is checked before the first rename; an existing archive is never replaced. Test through the exported `archiveOneSource` with a fixed stamp.
- The store and its sidecars move together or not at all: a rename that fails midway puts the files already moved back under their original names, then the error is reported on that source. Test: sidecar rename forced to fail, store back in place, no half-archive on disk.
- A source none of whose tables could be read is not archived; its entry says `archiveSkipped: no table could be merged from this source` and the file stays where the doctor can still see it. Test: a SQLite file with an unrelated table.

Suite: 16 tests in the file, full run green.

Second round, two more points taken:

- Identity is settled by device and inode (`fs.statSync` with bigint) on both files, which now catches a hard link to the live store as well; a filesystem that reports no inode falls back to the resolved names, case-folded on Windows. Both files exist by the time the check runs, so there is no unreachable fallback.
- Linked sources are refused before the merge: a symlink (`source is a symbolic link; pass the file it points to`) because archiving the link would leave its target for the doctor to flag, and a file with other hard links (`source has other hard links; archiving would not detach it`) because the archive would stay tied to whatever the other names write. Tests: symlink to a fragment on POSIX, hard link twin and hard link to the canonical on every platform.

Two Sonar findings on the first round fixed as well (`toReversed()` in the rollback, redundant empty-array check removed), plus tests for the failed rollback and the CLI error exit. 21 tests in the file.

Third round, from the full review:

- The canonical guard runs only when the canonical file exists on disk, so it no longer leans on the adapter persisting the store synchronously at open time (it does today, which is why the fresh-store dry run test passed; the guard removes the coupling).
- On a filesystem that reports no inode, the fallback compares real paths (links followed), not spellings.
- Archive destinations are checked with `lstat`, so a dangling symlink at the destination counts as taken instead of being replaced by the rename.
- With `--canonical :memory:` and `--apply`, nothing is archived: the in-memory store keeps nothing after close, so the sources stay the only copy.
- Two requests to skip archiving a copy whose rows are all already present were declined on purpose: that copy is exactly what the doctor must stop listing (the two files in #1380 were that case), and the archive is a rename next to the original.

22 tests in the file, full run green.

### CODE REVIEW

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 1]
```diff
@@ -103,13 +108,23 @@ function mergeOneTable(canonicalDb, srcDb, name, pk, apply) {
   };
 }
 
-async function mergeOneSource(canonicalDb, srcPath, apply) {
+function samePath(a, b) {
+  return path.resolve(a) === path.resolve(b);
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js]: Taken in a4f058aa: samePath now compares filesystem identity through fs.realpathSync.native on both sides, with a case-folded fallback on Windows when a path does not exist yet. A symlink to the live store and an upper-cased spelling of its name are both refused; covered by a test that uses a symlink on POSIX and the other spelling on Windows.

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 2]
```diff
@@ -119,15 +134,51 @@ async function mergeOneSource(canonicalDb, srcPath, apply) {
...
 }
 
-async function mergeStateDbs({ canonicalPath, sourcePaths, apply = false }) {
+// SQLite may leave a write-ahead log or a rollback journal next to a store;
+// they belong to the file they sit beside, so they move with it.
+const SIDECAR_SUFFIXES = ['-wal', '-shm', '-journal'];
+
+function archiveOneSource(srcPath, stamp) {
+  const archivedTo = `${srcPath}.merged-${stamp}.bak`;
+  fs.renameSync(srcPath, archivedTo);
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js]: Taken in a4f058aa: every destination (main file and each sidecar that will move) is checked before the first rename, and an existing archive makes the whole step fail with 'archive destination already exists' instead of replacing it. Tested through the exported archiveOneSource with a fixed stamp, for the main file and for a sidecar.

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 3]
```diff
@@ -119,15 +134,51 @@ async function mergeOneSource(canonicalDb, srcPath, apply) {
...
+  for (const suffix of SIDECAR_SUFFIXES) {
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js:182]: Taken in a4f058aa: the moves are recorded as they succeed and a failure midway renames them back in reverse order before the error is reported on that source. Test forces the sidecar rename to fail and asserts the store is back under its original name, the sidecar never moved, and no .merged- file is left in the directory.

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 4]
```diff
@@ -119,15 +134,51 @@ async function mergeOneSource(canonicalDb, srcPath, apply) {
...
+
+// Only after the canonical store is written, flushed and closed: a source
+// that failed to merge stays where it is, and a rename that fails (a file
+// held open by another process on Windows) is reported, not thrown, since
+// the merge itself already succeeded.
+function archiveSources(reports) {
+  const stamp = fileStamp();
+  const archived = [];
+  for (const report of reports) {
+    if (report.error) continue;
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js:223]: Taken in a4f058aa: a source none of whose tables could be read is not archived; its entry gets archiveSkipped with the reason and the file stays where the doctor can still see it. Sources where tables exist but hold nothing new (the real-world case that started this) are still archived, since nothing was left behind. Test uses a SQLite file with one unrelated table.

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 5]
```diff
@@ -119,15 +143,85 @@ async function mergeOneSource(canonicalDb, srcPath, apply) {
...
+  for (const suffix of SIDECAR_SUFFIXES) {
+    if (fs.existsSync(`${srcPath}${suffix}`)) moves.push([`${srcPath}${suffix}`, `${archivedTo}${suffix}`]);
+  }
+  for (const [, to] of moves) {
+    if (fs.existsSync(to)) throw new Error(`archive destination already exists: ${to}`);
+  }
+  const done = [];
+  try {
+    for (const [from, to] of moves) {
+      fs.renameSync(from, to);
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js:193]: Taken in 71063b71: a source that is a symlink is refused before the merge with 'source is a symbolic link; pass the file it points to', so neither the link nor its target ever moves. Test creates a symlink to a seeded fragment on POSIX and asserts the refusal and that both files stay.

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 6]
```diff
@@ -103,13 +108,32 @@ function mergeOneTable(canonicalDb, srcDb, name, pk, apply) {
...
 }
 
-async function mergeOneSource(canonicalDb, srcPath, apply) {
+// Filesystem identity, not spelling: a symlink to the canonical store, or the
+// same file spelled in another case on Windows, must be refused too.
+function samePath(a, b) {
+  const resolvedA = path.resolve(a);
+  const resolvedB = path.resolve(b);
+  try {
+    return fs.realpathSync.native(resolvedA) === fs.realpathSync.native(resolvedB);
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js]: Taken in 71063b71: identity is now settled by device and inode (fs.statSync with bigint) on both files, so a hard link to the live store is refused as the canonical store; a source that has other hard links is refused too, with 'source has other hard links; archiving would not detach it'. Test creates a hard-link twin of a fragment and a hard link to the canonical and asserts both refusals with every file left in place.

**Comment by Fmarzochi** [comment]
@cubic-dev-ai review

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 7]
```diff
@@ -119,15 +159,84 @@ async function mergeOneSource(canonicalDb, srcPath, apply) {
...
+    }
+    throw err;
+  }
+  return archivedTo;
+}
+
+// A file none of whose tables could be read is not a merged copy: renaming
+// it would only hide it from the doctor with everything still inside.
+function nothingMerged(report) {
+  return Object.values(report.tables).every(table => table.skipped);
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js:212]: Kept as is, on purpose. A copy whose rows are all already present in the canonical store is exactly the case this PR exists for: the doctor lists it, the person runs the merge, nothing is missing, and the copy has to stop being flagged. That was the real report in #1380 (two copies with one install_state row each, both already present). Leaving those in place would keep the warning forever. The archive is a rename next to the original, so nothing is lost either way. What must not be archived is a file the script could not read at all, and the all-skipped check covers that.

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 8]
```diff
@@ -119,15 +159,84 @@ async function mergeOneSource(canonicalDb, srcPath, apply) {
...
+// Only after the canonical store is written, flushed and closed: a source
+// that failed to merge stays where it is, and a rename that fails (a file
+// held open by another process on Windows) is reported, not thrown, since
+// the merge itself already succeeded.
+function archiveSources(reports) {
+  const stamp = fileStamp();
+  const archived = [];
+  for (const report of reports) {
+    if (report.error) continue;
+    if (nothingMerged(report)) {
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js:224]: Kept as is, same reason as the thread above: a readable copy whose rows are all already present is precisely the copy the doctor must stop listing (the two files in #1380 were exactly that). Requiring inserted > 0 would leave them flagged forever. The archive is a rename next to the original; nothing is lost. Only a file the script could not read at all stays in place, and that case is covered.

File: tests/scripts/merge-fragmented-state-dbs.test.js [hunk 9]
```diff
@@ -224,6 +261,383 @@ async function runTests() {
...
+  })) passed++; else failed++;
+
+  if (await test('a dry run against a canonical store that does not exist yet still tells the source apart from it', async () => {
+    const dir = createTempDir('egc-merge-fresh-');
+    try {
+      const canonicalPath = path.join(dir, 'fresh', 'state.db');
+      const sourcePath = path.join(dir, 'source.db');
+      await seedDb(sourcePath);
+
+      const result = await mergeStateDbs({ canonicalPath, sourcePaths: [sourcePath], apply: false });
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply tests/scripts/merge-fragmented-state-dbs.test.js:599]: The test passes on every run here and in the full suite: the store adapter persists synchronously on the first write (see the note above _persist in scripts/lib/state-store/db-adapter.js), so the canonical file exists before mergeOneSource runs. Taken anyway in 70f71003: the guard now runs only when the canonical file exists on disk, so it no longer leans on that adapter detail.

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 10]
```diff
@@ -103,13 +108,48 @@ function mergeOneTable(canonicalDb, srcDb, name, pk, apply) {
...
+// checked, the canonical store was just opened), so device and inode settle
+// it; a filesystem that reports no inode falls back to the resolved names,
+// case-folded on Windows.
+function sameFile(a, b) {
+  const statA = fs.statSync(a, { bigint: true });
+  const statB = fs.statSync(b, { bigint: true });
+  const fold = value => (process.platform === 'win32' ? value.toLowerCase() : value);
+  return statA.ino !== 0n && statB.ino !== 0n
+    ? statA.dev === statB.dev && statA.ino === statB.ino
+    : fold(path.resolve(a)) === fold(path.resolve(b));
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js]: Taken in 70f71003: without inodes the fallback compares fs.realpathSync.native on both sides (links followed, case-folded on Windows), and both files exist by then, so no try/catch is needed.

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 11]
```diff
@@ -119,15 +159,84 @@ async function mergeOneSource(canonicalDb, srcPath, apply) {
...
+// a rename that fails midway puts the files already moved back under their
+// original names before the error surfaces.
+function archiveOneSource(srcPath, stamp) {
+  const archivedTo = `${srcPath}.merged-${stamp}.bak`;
+  const moves = [[srcPath, archivedTo]];
+  for (const suffix of SIDECAR_SUFFIXES) {
+    if (fs.existsSync(`${srcPath}${suffix}`)) moves.push([`${srcPath}${suffix}`, `${archivedTo}${suffix}`]);
+  }
+  for (const [, to] of moves) {
+    if (fs.existsSync(to)) throw new Error(`archive destination already exists: ${to}`);
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js]: Taken in 70f71003: destinations are checked with fs.lstatSync(to, { throwIfNoEntry: false }), so a dangling symlink counts as taken. Test creates one at the destination on POSIX and asserts the refusal.

File: scripts/maintenance/merge-fragmented-state-dbs.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js]: Taken in 70f71003, same change as the earlier thread on this line: the no-inode fallback compares real paths, not spellings.

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 12]
```diff
@@ -137,35 +246,38 @@ async function mergeStateDbs({ canonicalPath, sourcePaths, apply = false }) {
...
   const commit = canonicalDb.transaction(() => {});
-  for (const src of sourcePaths) reports.push(await mergeOneSource(canonicalDb, src, apply));
+  for (const src of sourcePaths) reports.push(await mergeOneSource(canonicalDb, src, apply, resolvedCanonicalPath));
   if (apply) commit(); // no-op body; forces a single persist after all inserts above
 
   if (apply) await canonicalDb.flush();
   canonicalDb.close();
 
-  return { apply, canonical: resolvedCanonicalPath, backupPath, reports };
+  const archived = apply && !keepSources ? archiveSources(reports) : [];
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js]: Taken in 70f71003: with --canonical :memory: nothing is archived, whatever the other flags say. Test runs --apply against :memory: and asserts the source stays.

**Comment by Fmarzochi** [comment]
@cubic-dev-ai review

File: scripts/maintenance/merge-fragmented-state-dbs.js [hunk 13]
```diff
@@ -119,15 +157,86 @@ async function mergeOneSource(canonicalDb, srcPath, apply) {
...
+    }
+    throw err;
+  }
+  return archivedTo;
+}
+
+// A file none of whose tables could be read is not a merged copy: renaming
+// it would only hide it from the doctor with everything still inside.
+function nothingMerged(report) {
+  return Object.values(report.tables).every(table => table.skipped);
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply scripts/maintenance/merge-fragmented-state-dbs.js:212]: Declined for the third time, same reason: a readable copy whose rows are all already present is the copy the doctor must stop listing; that is the case from #1380 and the reason this PR exists. Zero inserts means nothing was left behind, and the archive is a rename next to the original. Only a file the script could not read at all stays in place.

File: tests/scripts/merge-fragmented-state-dbs.test.js [hunk 14]
```diff
@@ -5,14 +5,49 @@ const fs = require('fs');
...
 
-const { mergeStateDbs } = require('../../scripts/maintenance/merge-fragmented-state-dbs');
+const { execFileSync } = require('child_process');
+
+const { mergeStateDbs, archiveOneSource } = require('../../scripts/maintenance/merge-fragmented-state-dbs');
 const { openDatabase } = require('../../scripts/lib/state-store/db-adapter');
 const { applyMigrations } = require('../../scripts/lib/state-store/migrations');
 
+const DOCTOR_SCRIPT = path.join(__dirname, '..', '..', 'scripts', 'doctor.js');
+const CLI_TIMEOUT_MS = process.platform === 'win32' ? 30000 : 10000;
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply tests/scripts/merge-fragmented-state-dbs.test.js]: Taken in fda65124: both helpers take CLI_TIMEOUT_MS from tests/fixtures/subprocess-timeouts.js.
